src/common/common_deep_persist.c

```diff
@@ -0,0 +1,177 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * common_deep_persist.c -- internal utility functions for flushing
+ * a memory range.
+ * Currently only used on Linux.
+ */
+
+#define _GNU_SOURCE
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "os.h"
+#include "file.h"
+#include "out.h"
+#include "common_deep_persist.h"
+#include "os_deep_persist.h"
+#include "libpmem.h"
+
+/*
+ * ddax_deep_flush_write -- perform final write to deep_flush file
+ * on given region_id
+ */
+int
+ddax_deep_flush_write(int region_id)
+{
+	LOG(3, "region_id %d", region_id);
+
+	char deep_flush_path[PATH_MAX];
+	int deep_flush_fd;
+
+	snprintf(deep_flush_path, PATH_MAX,
+		"/sys/bus/nd/devices/region%d/deep_flush", region_id);
+
+	if ((deep_flush_fd = os_open(deep_flush_path, O_RDWR)) < 0) {
+		ERR("!os_open(\"%s\", O_RDWR)", deep_flush_path);
+		return -1;
+	}
+
+	if (write(deep_flush_fd, "1", 1) != 1) {
+		ERR("!write(%d, \"1\")", deep_flush_fd);
+		int oerrno = errno;
+		os_close(deep_flush_fd);
+		errno = oerrno;
+		return -1;
+	}
+
+	os_close(deep_flush_fd);
+	return 0;
+}
+
+/*
+ * common_replica_deep_persist -- perform deep persist on replica's parts
+ * for a given range. For dev dax write to deep_flush file from sysfs,
+ * otherwise call msync
+ */
+static int
+common_replica_deep_persist(const void *addr, size_t len,
+			struct pool_set *set, unsigned replica_id)
```

Comment:
can't you just have "struct pool_replica *rep" argument instead of pool_set and replica_id?